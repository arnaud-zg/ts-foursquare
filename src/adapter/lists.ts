import { NLists, NPayload } from '../../types'
import { adaptPayload } from './payload'
import { ECommonError } from '../constants/error'

export const adaptGetLists = (
  payload: NPayload.IPayload<NLists.IResponse>
): NLists.IList | null => {
  const response = adaptPayload<NLists.IResponse>(payload)

  if (!response || !response.list) {
    throw new Error(ECommonError.RESPONSE_NOT_VALID)
  }

  return response.list
}
