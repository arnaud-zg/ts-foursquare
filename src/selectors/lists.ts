import { NLists, NStore } from '../../types'

export const listSelector = (
  state: NStore.IState,
  listId: string
): NLists.IList => listsSelector(state).entities[listId]

export const listsSelector = (state: NStore.IState) => state.lists
