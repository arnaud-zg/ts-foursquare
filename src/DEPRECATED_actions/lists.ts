import { ActionType, createAsyncAction } from 'typesafe-actions'
import { NRequest } from '../../types'
import { adaptGetLists } from '../adapter'

export const getListsAsync = createAsyncAction(
  'GET_LISTS_REQUEST',
  'GET_LISTS_SUCCESS',
  'GET_LISTS_FAILURE',
  'GET_LISTS_CANCEL'
)<NRequest.IListsPayload, ReturnType<typeof adaptGetLists>, Error, string>()

export const listsActions = {
  getListsAsync,
}

export type TListsAction = ActionType<typeof listsActions>
