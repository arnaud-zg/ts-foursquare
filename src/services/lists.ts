import { fromFetch } from 'rxjs/fetch'
import { catchError, switchMap, map } from 'rxjs/operators'
import { NRequest } from '../../types'
import { EApiDefaultParameters, EApiPathnames } from '../constants/api'
import { generatePath } from '../utils/generatePath'
import { getLocationHref } from '../utils/url'
import { processFetchError, processFetchResponse } from './fetch'
import { adaptGetLists } from '../adapter'
import { IStandaloneConfig } from '../standalone'

export const getObservableLists = ({
  payload,
}: {
  config?: IStandaloneConfig
  payload: NRequest.IListsPayload
}) => {
  const { listId } = payload

  return fromFetch(
    getLocationHref({
      origin: EApiDefaultParameters.ORIGIN,
      pathname: generatePath(EApiPathnames.LISTS, { listId }),
      param: {},
    })
  ).pipe(
    switchMap(processFetchResponse),
    map(adaptGetLists),
    catchError(processFetchError)
  )
}
