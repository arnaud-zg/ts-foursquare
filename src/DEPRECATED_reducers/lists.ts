import { createReducer } from 'typesafe-actions'
import { NListsState } from '../../types'
import { getListsAsync, TListsAction } from '../DEPRECATED_actions'

export const initialState: NListsState.IState = {
  entities: {},
}

export const listsReducer = createReducer<NListsState.IState, TListsAction>(
  initialState
).handleAction(
  getListsAsync.success,
  (state, action): NListsState.IState => {
    return {
      ...state,
      entities: {
        ...state.entities,
        [action.payload.id]: action.payload,
      },
    }
  }
)
