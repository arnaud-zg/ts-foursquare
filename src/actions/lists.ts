import { NRequest } from '../../types'
import { getObservableLists } from '../services/lists'
import { IStandaloneConfig } from '../standalone'

export interface GetListsProps {
  config?: IStandaloneConfig
  payload: NRequest.IListsPayload
}

export const getLists = ({ config, payload }: GetListsProps) =>
  getObservableLists({ config, payload }).toPromise()
