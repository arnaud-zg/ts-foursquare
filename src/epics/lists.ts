import { Epic } from 'redux-observable'
import { of } from 'rxjs/internal/observable/of'
import { catchError, filter, map, switchMap, takeUntil } from 'rxjs/operators'
import { isActionOf } from 'typesafe-actions'
import { NStore } from '../../types'
import { getListsAsync, TRootAction } from '../actions'
import { adaptGetLists } from '../adapter'
import { getObservableLists } from '../services'

export const getListsEpic: Epic<TRootAction, TRootAction, NStore.IState> = (
  action$,
  state$
) =>
  action$.pipe(
    filter(isActionOf(getListsAsync.request)),
    switchMap(action =>
      getObservableLists({ action, state$ }).pipe(
        map(adaptGetLists),
        map(getListsAsync.success),
        catchError(err => of(getListsAsync.failure(err))),
        takeUntil(action$.pipe(filter(isActionOf(getListsAsync.cancel))))
      )
    )
  )
