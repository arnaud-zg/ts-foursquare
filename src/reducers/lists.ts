import { createReducer } from 'typesafe-actions'
import { NListsState } from '../../types'
import { getListsAsync, TListsAction } from '../actions'

export const initialState: NListsState.IState = {
  entities: {},
}

export const listsReducer = createReducer<NListsState.IState, TListsAction>(
  initialState
).handleAction(
  getListsAsync.success,
  (state, action): NListsState.IState => {
    return {
      ...state,
      entities: {
        ...state.entities,
        [action.payload.id]: action.payload,
      },
    }
  }
)
