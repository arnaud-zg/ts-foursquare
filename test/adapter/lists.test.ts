import { adaptGetLists } from '../../src/adapter/lists'
import { payload as payloadGetLists } from '../actions/__mocks__/getLists.resolve'

describe('adapter/lists/adaptGetLists', () => {
  it('should get an empty list', () => {
    expect(() =>
      adaptGetLists({
        ...payloadGetLists,
        meta: {
          code: 400,
          requestId: payloadGetLists.meta.requestId,
        },
      })
    ).toThrowErrorMatchingSnapshot()
  })

  it('should get a list of categories', () => {
    expect(adaptGetLists(payloadGetLists)).toMatchSnapshot()
  })
})
