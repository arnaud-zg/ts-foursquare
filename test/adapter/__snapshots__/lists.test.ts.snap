// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adapter/lists/adaptGetLists should get a list of categories 1`] = `
Object {
  "canonicalUrl": "https://foursquare.com/foursquare/list/nycs-best-rooftops",
  "categories": Object {
    "count": 9,
    "items": Array [
      Object {
        "category": Object {
          "icon": Object {
            "prefix": "https://ss3.4sqi.net/img/categories_v2/nightlife/pub_",
            "suffix": ".png",
          },
          "id": "4bf58dd8d48988d116941735",
          "name": "Bar",
          "pluralName": "Bars",
          "shortName": "Bar",
        },
        "count": 8,
      },
      Object {
        "category": Object {
          "icon": Object {
            "prefix": "https://ss3.4sqi.net/img/categories_v2/travel/hotel_",
            "suffix": ".png",
          },
          "id": "4bf58dd8d48988d1fa931735",
          "name": "Hotel",
          "pluralName": "Hotels",
          "shortName": "Hotel",
        },
        "count": 4,
      },
    ],
  },
  "collaborative": false,
  "collaborators": Object {
    "count": 0,
    "items": Array [],
  },
  "completedCount": 0,
  "createdAt": 1434481182,
  "description": "",
  "editable": false,
  "followers": Object {
    "count": 350,
  },
  "following": false,
  "id": "5580721e498e7c48540bf83f",
  "listItems": Object {
    "count": 10,
    "items": Array [
      Object {
        "createdAt": 1493226584,
        "id": "t57e71351498e9292956226a1",
        "index": 0,
        "isNew": true,
        "tip": Object {
          "agreeCount": 3,
          "canonicalUrl": "https://foursquare.com/item/57e71351498e9292956226a1",
          "createdAt": 1474761553,
          "disagreeCount": 0,
          "id": "57e71351498e9292956226a1",
          "logView": true,
          "saves": Object {},
          "text": "Awesome new rooftop with relatively cheap beer and great views of NYC. Get there right at 4pm when they open or be prepared to wait in line.",
          "todo": Object {
            "count": 17,
          },
          "type": "user",
          "user": Object {},
        },
        "venue": Object {
          "allowMenuUrlEdit": true,
          "beenHere": Object {
            "lastCheckinExpiredAt": 0,
          },
          "contact": Object {
            "formattedPhone": "(718) 307-7100",
            "phone": "7183077100",
          },
          "id": "57acf9af498e6b8d5376c2d4",
          "location": Object {
            "address": "111 N 12th St",
            "cc": "US",
            "city": "Brooklyn",
            "country": "United States",
            "crossStreet": "at Wythe Ave",
            "formattedAddress": Array [
              "111 N 12th St (at Wythe Ave)",
              "Brooklyn, NY 11249",
              "United States",
            ],
            "labeledLatLngs": Array [
              Object {
                "label": "display",
                "lat": 40.722294869596624,
                "lng": -73.95677447319031,
              },
            ],
            "lat": 40.722294869596624,
            "lng": -73.95677447319031,
            "postalCode": "11249",
            "state": "NY",
          },
          "name": "Westlight",
          "rating": 9.1,
          "ratingColor": "00B551",
          "ratingSignals": 353,
          "stats": Object {
            "checkinsCount": 3253,
            "tipCount": 62,
            "usersCount": 2517,
          },
          "url": "http://westlightnyc.com",
          "verified": false,
        },
      },
    ],
  },
  "logView": true,
  "name": "NYC's Best Rooftops",
  "photo": Object {
    "createdAt": 1411429238,
    "height": 720,
    "id": "5420b376498e3fdb42d08c8d",
    "prefix": "https://igx.4sqi.net/img/general/",
    "suffix": "/18091_SPs8d1c-WsbNGV89cUg9ou1QOGa7YcDlRJLUlh1b6CQ.jpg",
    "visibility": "public",
    "width": 960,
  },
  "public": true,
  "sort": "list-order",
  "type": "others",
  "updatedAt": 1497977804,
  "user": Object {
    "firstName": "Foursquare City Guide",
    "gender": "none",
    "id": "23438729",
    "photo": Object {
      "prefix": "https://igx.4sqi.net/img/user/",
      "suffix": "/23438729-TODURDIUDUMY4JF5.png",
    },
    "type": "page",
  },
  "venueCount": 10,
  "visitedCount": 0,
}
`;

exports[`adapter/lists/adaptGetLists should get an empty list 1`] = `"Response is not valid"`;
